File: dbfit-dotnet/sqlserver/SqlServerEnvironment.cs
/// Copyright (C) Gojko Adzic 2006-2008 http://gojko.net
/// Released under GNU GPL 2.0

using System;
using System.Collections.Generic;
using System.Data;
using System.Data.Common;
using System.Data.SqlClient;
using System.Text.RegularExpressions;
using System.Xml;
using System.Text;
using dbfit.util;

namespace dbfit
{
    /// <summary>
    /// Implementation of IDbEnvironment that works with SqlServer 2005 and newer versions
    /// </summary>
    public class SqlServerEnvironment : AbstractDbEnvironment
    {
        private static readonly int MAX_STRING_SIZE = 4000;
        protected override String GetConnectionString(String dataSource, String username, String password, String databaseName)
        {
            return String.Format("data source={0};user id={1};password={2};database={3};", dataSource, username, password, databaseName);
        }

        protected override String GetConnectionString(String dataSource, String username, String password)
        {
            return String.Format("Data Source={0}; User ID={1}; Password={2}", dataSource, username, password);
        }
        private static DbProviderFactory dbp = DbProviderFactories.GetFactory("System.Data.SqlClient");
        private Regex paramNames = new Regex("@([A-Za-z0-9_]*)");
        protected override Regex ParamNameRegex { get { return paramNames;}}

        public override DbProviderFactory DbProviderFactory
        {
            get { return dbp; }
        }
        public override Dictionary<String, DbParameterAccessor> GetAllProcedureParameters(String procName)
        {
            
            return ReadIntoParams(procName,
               @"select p.[name], TYPE_NAME(p.system_type_id) as [Type], 
			   p.max_length, p.is_output, p.is_cursor_ref, p.precision, p.scale from sys.parameters p 
				where p.object_id = OBJECT_ID(@objname) order by parameter_id"
               );
        }
        public override Dictionary<String, DbParameterAccessor> GetAllColumns(String tableOrViewName)
        {

            return ReadIntoParams(tableOrViewName,
            @"select c.[name], TYPE_NAME(c.system_type_id) as [Type], c.max_length, 
            0 As is_output, 0 As is_cursor_ref, c.precision, c.scale
            from sys.columns c
            where c.object_id = OBJECT_ID(@objname)
            order by column_id
            " );
        }

        private  Dictionary<string, DbParameterAccessor> ReadIntoParams(String objname, String query)
        {
            if (objname.Contains("."))
            {
                String[] schemaAndName = objname.Split(new char[] { '.' }, 2);
                objname = "[" + schemaAndName[0] + "].[" + schemaAndName[1] + "]";
            }
            else
            {
                objname = "[" + NameNormaliser.NormaliseName(objname) + "]";
            }
            DbCommand dc = CurrentConnection.CreateCommand();
            dc.Transaction = CurrentTransaction;
            dc.CommandText = query;
            dc.CommandType = CommandType.Text;
            AddInput(dc, "@objname", objname);
            DbDataReader reader = dc.ExecuteReader();
            Dictionary<String, DbParameterAccessor>
                allParams = new Dictionary<string, DbParameterAccessor>();
            int position=0;
            while (reader.Read())
            {

                String paramName = (reader.IsDBNull(0)) ? null : reader.GetString(0);
                String dataType = reader.GetString(1);
                int length = (reader.IsDBNull(2)) ? 0 : System.Convert.ToInt32(reader[2]);
                int isOutput = (reader.IsDBNull(3)) ? 0 : System.Convert.ToInt32(reader[3]);
                byte precision =  System.Convert.ToByte(reader[5]);
                byte scale = System.Convert.ToByte(reader[6]);

                SqlParameter dp = new SqlParameter();
                dp.Direction = GetParameterDirection(isOutput);
                if (!String.IsNullOrEmpty(paramName)) { 
						dp.ParameterName = paramName; dp.SourceColumn=paramName; 
				}
                else
                {
                    dp.Direction = ParameterDirection.ReturnValue;
                }
                dp.SqlDbType= GetDBType(dataType);
                String typeName = NormaliseTypeName(dataType);
                if (precision > 0) dp.Precision = precision;         
                if (scale > 0) dp.Scale = scale;
                if ("NTEXT".Equals(typeName)||("TEXT".Equals(typeName)))
                    dp.Size=MAX_STRING_SIZE;
                else if (length > 0)
                {
                    dp.Size = System.Convert.ToInt32(length);
                }
                else
                {
                    if (!ParameterDirection.Input.Equals(dp.Direction) || 
                        typeof(String).Equals(GetDotNetType(dataType)))
                        dp.Size = MAX_STRING_SIZE;
                }
                allParams[NameNormaliser.NormaliseName(paramName)] =
                    new DbParameterAccessor(dp, GetDotNetType(dataType), position++, dataType);
            }
            reader.Close();
            if (allParams.Count == 0)
                throw new ApplicationException("Cannot read columns/parameters for object " + objname + " - check spelling or access privileges ");
            return allParams;
        }
        private static string[] StringTypes = new string[] { "VARCHAR", "NVARCHAR", "CHAR", "NCHAR","TEXT","NTEXT","XML"};
        private static string[] DecimalTypes = new string[] { "DECIMAL", "NUMERIC", "MONEY", "SMALLMONEY" };
        private static string[] DateTypes = new string[] { "SMALLDATETIME","DATETIME","TIMESTAMP" };
        private static string[] RefCursorTypes = new string[] { "REF" };
        private static string[] Int32Types=new string[] {"INT"};
        private static string[] Int16Types = new string[] { "TINYINT", "SMALLINT"};
        private static string[] Int64Types = new string[] { "BIGINT"};

        private static string[] BooleanTypes = new string[] { "BIT" };
		private static string[] BinaryTypes=new string[] {"BINARY","VARBINARY","IMAGE"};
        private static string[] GuidTypes = new string[] { "UNIQUEIDENTIFIER" };
        private static string[] VariantTypes = new string[] { "SQL_VARIANT" };
        private static string[] FloatTypes = new String[] { "REAL", "FLOAT" };
        private static string NormaliseTypeName(string dataType)
        {
            dataType = dataType.ToUpper().Trim();
            int idx = dataType.IndexOf(" ");
            if (idx >= 0) dataType = dataType.Substring(0, idx);
            idx = dataType.IndexOf("(");
            if (idx >= 0) dataType = dataType.Substring(0, idx);
            return dataType;
            
        }
        protected static SqlDbType GetDBType(String dataType)
        { 
            //todo:strip everything from first blank
            dataType = NormaliseTypeName(dataType);

            if (Array.IndexOf(StringTypes, dataType) >= 0) return SqlDbType.VarChar;
            if (Array.IndexOf(DecimalTypes, dataType) >= 0) return SqlDbType.Decimal;
            if (Array.IndexOf(DateTypes, dataType) >= 0) return SqlDbType.DateTime;
            if (Array.IndexOf(Int32Types, dataType) >= 0) return SqlDbType.Int;
            if (Array.IndexOf(Int16Types, dataType) >= 0) return SqlDbType.Int;
            if (Array.IndexOf(Int64Types, dataType) >= 0) return SqlDbType.Int;
            if (Array.IndexOf(BooleanTypes, dataType) >= 0) return SqlDbType.Bit;
			if (Array.IndexOf(BinaryTypes,dataType)>=0) return SqlDbType.VarBinary;
            //if (Array.IndexOf(RefCursorTypes, dataType) >= 0) return OracleType.Cursor;
            if (Array.IndexOf(GuidTypes, dataType) >= 0) return SqlDbType.UniqueIdentifier;
            if (Array.IndexOf(VariantTypes, dataType) >= 0) return SqlDbType.Variant;
            if (Array.IndexOf(FloatTypes, dataType) >= 0) return SqlDbType.Float;

            throw new NotSupportedException("Type " + dataType + " is not supported");
        }
        protected static Type GetDotNetType(String dataType)
        {
            dataType = NormaliseTypeName(dataType);
            if (Array.IndexOf(StringTypes, dataType) >= 0) return typeof(string);
            if (Array.IndexOf(DecimalTypes, dataType) >= 0) return typeof(decimal);
            if (Array.IndexOf(Int32Types, dataType) >= 0) return typeof(Int32);
            if (Array.IndexOf(Int16Types, dataType) >= 0) return typeof(Int16);
            if (Array.IndexOf(Int64Types, dataType) >= 0) return typeof(Int64);
            if (Array.IndexOf(DateTypes, dataType) >= 0) return typeof(DateTime);
            if (Array.IndexOf(RefCursorTypes, dataType) >= 0) return typeof(DataTable);
            if (Array.IndexOf(BooleanTypes, dataType) >= 0) return typeof(bool);
			if (Array.IndexOf(BinaryTypes, dataType) >= 0) return typeof(byte[]);
            if (Array.IndexOf(GuidTypes, dataType) >= 0) return typeof(System.Guid);
            if (Array.IndexOf(VariantTypes, dataType) >= 0) return typeof(string);
            if (Array.IndexOf(FloatTypes, dataType) >= 0) return typeof(double);

            throw new NotSupportedException("Type " + dataType + " is not supported");
        }
        private static ParameterDirection GetParameterDirection(int isOutput)
        {
            if (isOutput==1) return ParameterDirection.Output;
            else return ParameterDirection.Input;
        }
        public override bool SupportsReturnOnInsert { get { return false; } }
        public override String IdentitySelectStatement { get { return "select @@identity"; } }

        public override int GetExceptionCode(Exception dbException)
        {

            if (dbException is System.Data.SqlClient.SqlException)
            {
                Console.WriteLine("SQL Exception " + ((System.Data.SqlClient.SqlException)dbException).Number);
                return ((System.Data.SqlClient.SqlException)dbException).Number;
            }
            else return base.GetExceptionCode(dbException);
        }
		 public override string ParameterPrefix
		 {
			 get { return "@";}
		 }
    }
}

File: dbfit-java/sqlserver/src/main/java/dbfit/environment/SqlServerEnvironment.java
package dbfit.environment;

import dbfit.annotations.DatabaseEnvironment;
import dbfit.api.AbstractDbEnvironment;
import dbfit.util.DbParameterAccessor;
import dbfit.util.DbParameterAccessorsMapBuilder;
import dbfit.util.Direction;
import static dbfit.util.Direction.*;
import static dbfit.util.LangUtils.enquoteAndJoin;
import dbfit.util.TypeNormaliserFactory;
import static dbfit.environment.SqlServerTypeNameNormaliser.normaliseTypeName;

import java.math.BigDecimal;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.regex.Pattern;
import java.util.Properties;

import static org.apache.commons.lang3.ObjectUtils.defaultIfNull;

@DatabaseEnvironment(name="SqlServer", driver="com.microsoft.sqlserver.jdbc.SQLServerDriver")
public class SqlServerEnvironment extends AbstractDbEnvironment {

    public SqlServerEnvironment(String driverClassName) {
        super(driverClassName);

        TypeNormaliserFactory.setNormaliser(java.sql.Time.class,
                new MillisecondTimeNormaliser());
    }

    public boolean supportsOuputOnInsert() {
        return false;
    }

    @Override
    protected String getConnectionString(String dataSource) {
        return "jdbc:sqlserver://" + dataSource;
    }

    @Override
    protected String getConnectionString(String dataSource, String database) {
        return getConnectionString(dataSource) + ";database=" + database;
    }

    @Override
    public void connect(String connectionString, Properties info) throws SQLException {
        // Add sendTimeAsDatetime=false option to enforce sending Time as
        // java.sql.Time (otherwise some precision is lost in conversions)
        super.connect(connectionString + ";sendTimeAsDatetime=false", info);
    }

    private static String paramNamePattern = "@([A-Za-z0-9_]+)";
    private static Pattern paramRegex = Pattern.compile(paramNamePattern);

    public Pattern getParameterPattern() {
        return paramRegex;
    }

    protected String parseCommandText(String commandText) {
        commandText = commandText.replaceAll(paramNamePattern, "?");
        return super.parseCommandText(commandText);
    }

    public Map<String, DbParameterAccessor> getAllColumns(String tableOrViewName)
            throws SQLException {
        String qry = " select c.[name], TYPE_NAME(c.system_type_id) as [Type], c.max_length, "
                + " 0 As is_output, 0 As is_cursor_ref "
                + " from " + objectDatabasePrefix(tableOrViewName) + "sys.columns c "
                + " where c.object_id = OBJECT_ID(?) "
                + " order by column_id";
        return readIntoParams(tableOrViewName, qry);
    }

    private Map<String, DbParameterAccessor> readIntoParams(String objname,
            String query) throws SQLException {
        DbParameterAccessorsMapBuilder params = new DbParameterAccessorsMapBuilder();

        objname = objname.replaceAll("[^a-zA-Z0-9_.#$]", "");
        String bracketedName = enquoteAndJoin(objname.split("\\."), ".", "[", "]");

        try (PreparedStatement dc = currentConnection.prepareStatement(query)) {
            dc.setString(1, bracketedName);
            ResultSet rs = dc.executeQuery();

            while (rs.next()) {
                String paramName = defaultIfNull(rs.getString(1), "");
                params.add(paramName,
                           getParameterDirection(rs.getInt(4), paramName),
                           getSqlType(rs.getString(2)),
                           getJavaClass(rs.getString(2)));
            }
        }

        return params.toMap();
    }

    // List interface has sequential search, so using list instead of array to
    // map types
    private static List<String> stringTypes = Arrays.asList(new String[] {
            "VARCHAR", "NVARCHAR", "CHAR", "NCHAR", "TEXT", "NTEXT",
            "UNIQUEIDENTIFIER" });
    private static List<String> intTypes = Arrays
            .asList(new String[] { "INT" });
    private static List<String> booleanTypes = Arrays
            .asList(new String[] { "BIT" });
    private static List<String> floatTypes = Arrays
            .asList(new String[] { "REAL" });
    private static List<String> doubleTypes = Arrays
            .asList(new String[] { "FLOAT" });
    private static List<String> longTypes = Arrays
            .asList(new String[] { "BIGINT" });
    private static List<String> shortTypes = Arrays.asList(new String[] {
            "TINYINT", "SMALLINT" });

    private static List<String> numericTypes = Arrays.asList("NUMERIC");
    private static List<String> decimalTypes = Arrays.asList(new String[] {
            "DECIMAL", "MONEY", "SMALLMONEY" });
    private static List<String> timestampTypes = Arrays.asList(new String[] {
            "SMALLDATETIME", "DATETIME", "DATETIME2" });
    private static List<String> dateTypes = Arrays.asList("DATE");
    private static List<String> timeTypes = Arrays.asList("TIME");

    // private static List<String> refCursorTypes = Arrays.asList(new String[] {
    // });
    // private static List<String> doubleTypes=Arrays.asList(new
    // String[]{"DOUBLE"});

    // private static string[] BinaryTypes=new string[] {"BINARY","VARBINARY", "TIMESTAMP"};
    // private static string[] GuidTypes = new string[] { "UNIQUEIDENTIFIER" };
    // private static string[] VariantTypes = new string[] { "SQL_VARIANT" };

    private String objectDatabasePrefix(String dbObjectName) {
        String objectDatabasePrefix = "";
        String[] objnameParts = dbObjectName.split("\\.");
        if (objnameParts.length == 3) {
        	objectDatabasePrefix = objnameParts[0] + ".";
        }
        return objectDatabasePrefix;
    }

    private static Direction getParameterDirection(int isOutput, String name) {
        if (name.isEmpty()) {
            return RETURN_VALUE;
        }

        return (isOutput == 1) ? OUTPUT : INPUT;
    }

    private static int getSqlType(String dataType) {
        // todo:strip everything from first blank
        dataType = normaliseTypeName(dataType);

        if (stringTypes.contains(dataType))
            return java.sql.Types.VARCHAR;
        if (numericTypes.contains(dataType))
            return java.sql.Types.NUMERIC;
        if (decimalTypes.contains(dataType))
            return java.sql.Types.DECIMAL;
        if (intTypes.contains(dataType))
            return java.sql.Types.INTEGER;
        if (timestampTypes.contains(dataType))
            return java.sql.Types.TIMESTAMP;
        if (dateTypes.contains(dataType))
            return java.sql.Types.DATE;
        if (timeTypes.contains(dataType))
            return java.sql.Types.TIME;
        if (booleanTypes.contains(dataType))
            return java.sql.Types.BOOLEAN;
        if (floatTypes.contains(dataType))
            return java.sql.Types.FLOAT;
        if (doubleTypes.contains(dataType))
            return java.sql.Types.DOUBLE;

        if (longTypes.contains(dataType))
            return java.sql.Types.BIGINT;
        if (shortTypes.contains(dataType))
            return java.sql.Types.SMALLINT;

        throw new UnsupportedOperationException("Type " + dataType
                + " is not supported");
    }

    public Class<?> getJavaClass(String dataType) {
        dataType = normaliseTypeName(dataType);
        if (stringTypes.contains(dataType))
            return String.class;
        if (numericTypes.contains(dataType))
            return BigDecimal.class;
        if (decimalTypes.contains(dataType))
            return BigDecimal.class;
        if (intTypes.contains(dataType))
            return Integer.class;
        if (timestampTypes.contains(dataType))
            return java.sql.Timestamp.class;
        if (dateTypes.contains(dataType))
            return java.sql.Date.class;
        if (timeTypes.contains(dataType))
            return java.sql.Time.class;
        if (booleanTypes.contains(dataType))
            return Boolean.class;
        if (floatTypes.contains(dataType))
            return Float.class;
        if (doubleTypes.contains(dataType))
            return Double.class;
        if (longTypes.contains(dataType))
            return Long.class;
        if (shortTypes.contains(dataType))
            return Short.class;

        throw new UnsupportedOperationException("Type " + dataType
                + " is not supported");
    }

    public Map<String, DbParameterAccessor> getAllProcedureParameters(
            String procName) throws SQLException {
        return readIntoParams(
                procName,
                "select [name], [Type], max_length, is_output, is_cursor_ref from "
                    + "("
                    + "   select "
                    + "       p.[name], TYPE_NAME(p.system_type_id) as [Type], "
                    + "       p.max_length, p.is_output, p.is_cursor_ref, "
                    + "       p.parameter_id, 0 as set_id, p.object_id "
                    + "   from " + objectDatabasePrefix(procName) + "sys.parameters p "
                    + "   union all select "
                    + "        '' as [name], 'int' as [Type], "
                    + "        4 as max_length, 1 as is_output, 0 as is_cursor_ref, "
                    + "        null as parameter_id, 1 as set_id, object_id "
                    + "   from " + objectDatabasePrefix(procName) + "sys.objects where type in (N'P', N'PC') "
                    + ") as u where object_id = OBJECT_ID(?) order by set_id, parameter_id");
    }

    public String buildInsertCommand(String tableName,
            DbParameterAccessor[] accessors) {
        StringBuilder sb = new StringBuilder("insert into ");
        sb.append(tableName).append("(");
        String comma = "";

        StringBuilder values = new StringBuilder();

        for (DbParameterAccessor accessor : accessors) {
            if (accessor.hasDirection(Direction.INPUT)) {
                sb.append(comma);
                values.append(comma);
                //This will allow column names that have spaces or are keywords.
                sb.append("[" + accessor.getName() + "]");
                values.append("?");
                comma = ",";
            }
        }
        sb.append(") values (");
        sb.append(values);
        sb.append(")");
        return sb.toString();
    }
}

